### PR TITLE
Upgrade frontend

### PR DESCRIPTION
**Pull request by elimisteve**

### CODE REVIEW

**Comment by elimisteve** [comment]
@jimmcgaw Created this as a draft PR just so it's much harder to lose track of among the other branches

**Comment by jimmcgaw** [comment]
> @jimmcgaw Created this as a draft PR just so it's much harder to lose track of among the other branches

Coolio. Removed the cruft and just left the frontend/ dir.

File: webpack.config.base.js [hunk 1]
```diff
@@ -1,8 +1,7 @@
 const path = require('path');
-const ExtractTextPlugin = require('extract-text-webpack-plugin');
+const MiniCssExtractPlugin = require("mini-css-extract-plugin");
```
Comment (jimmcgaw) [inline webpack.config.base.js:2]: This replaces what's just above it.

File: webpack.config.base.js [hunk 2]
```diff
@@ -1,8 +1,7 @@
...
 const HtmlWebpackPlugin = require('html-webpack-plugin');
-const CleanWebpackPlugin = require('clean-webpack-plugin');
```
Comment (jimmcgaw) [inline webpack.config.base.js:4]: Deprecated, now just `clean: true` in one of the config options.

File: webpack.config.base.js [hunk 3]
```diff
@@ -1,8 +1,7 @@
...
-const CopyWebpackPlugin = require('copy-webpack-plugin');
+const CopyPlugin = require('copy-webpack-plugin');
```
Comment (jimmcgaw) [inline webpack.config.base.js:4]: I don't remember why this var name changed during the process, not really necessary, but no harm done either.

File: webpack.config.base.js [hunk 4]
```diff
@@ -26,31 +26,43 @@ module.exports = {
       },
       {
         test: /\.(eot|svg|ttf|woff|woff2)$/,
-        loader: 'file-loader?name=[name].[ext]'
+        loader: 'file-loader',
+        options: {
+          'name': '[name].[ext]'
+        }
```
Comment (jimmcgaw) [inline webpack.config.base.js:32]: Query string options for the loader now specified as objects.

File: webpack.config.base.js [hunk 5]
```diff
@@ -26,31 +26,43 @@ module.exports = {
...
   resolve: {
     extensions: ['.js', '.jsx', '.css', '.scss', '.json'],
     modules: [
       'node_modules'
-    ]
+    ],
+    fallback: {
+      "crypto": require.resolve('crypto-browserify'),
+      "stream": require.resolve("stream-browserify"),
+      buffer: require.resolve("buffer/"),
```
Comment (jimmcgaw) [inline webpack.config.base.js:51]: Polyfills that were automatically brought in in webpack 4, but need to be specified in 5.

File: .eslintrc [hunk 6]
```diff
@@ -24,14 +24,13 @@
...
+      "experimentalObjectRestSpread": true,
     }
   },
   "rules": {
     "indent": ["error", 2],
-    "react/jsx-indent-props":  ["error", 2],
-    "react/jsx-indent": ["error", 2],
-    "semi": [2]
+    // "react/jsx-indent-props":  ["error", 2],
+    // "react/jsx-indent": ["error", 2],
```
Comment (jimmcgaw) [inline .eslintrc:33]: Another issue: the linter won't check the JSX indentation until I sort this out.

Also looks like I clobbered the `semi` rule, I'll at least add that back.

**Comment by elimisteve** [comment]
Looking great!

I do see this:

```js
WARNING in ./src/components/chat/MentionSuggestions.js 27:100-115
export 'ViewingUserIcon' (imported as 'ViewingUserIcon') was not found in './UserStatusIcons' (possible exports: UserStatusIcon)
 @ ./src/components/chat/AutoSuggest.js 5:0-54 18:130-148
 @ ./src/components/chat/ChatContainer.js 20:0-40 53:74-85
 @ ./src/components/App.js 22:0-49 136:43-56
 @ ./src/index.js 15:0-35 31:36-39

WARNING in ./src/components/chat/MentionSuggestions.js 27:167-181
export 'OnlineUserIcon' (imported as 'OnlineUserIcon') was not found in './UserStatusIcons' (possible exports: UserStatusIcon)
 @ ./src/components/chat/AutoSuggest.js 5:0-54 18:130-148
 @ ./src/components/chat/ChatContainer.js 20:0-40 53:74-85
 @ ./src/components/App.js 22:0-49 136:43-56
 @ ./src/index.js 15:0-35 31:36-39

WARNING in ./src/components/chat/MentionSuggestions.js 27:224-239
export 'OfflineUserIcon' (imported as 'OfflineUserIcon') was not found in './UserStatusIcons' (possible exports: UserStatusIcon)
 @ ./src/components/chat/AutoSuggest.js 5:0-54 18:130-148
 @ ./src/components/chat/ChatContainer.js 20:0-40 53:74-85
 @ ./src/components/App.js 22:0-49 136:43-56
 @ ./src/index.js 15:0-35 31:36-39
```

Beyond that this looks pretty great to merge!

**Comment by jimmcgaw** [comment]
> Looking great!
> 
> I do see this:
> 
> ```js
> WARNING in ./src/components/chat/MentionSuggestions.js 27:100-115
> export 'ViewingUserIcon' (imported as 'ViewingUserIcon') was not found in './UserStatusIcons' (possible exports: UserStatusIcon)
>  @ ./src/components/chat/AutoSuggest.js 5:0-54 18:130-148
>  @ ./src/components/chat/ChatContainer.js 20:0-40 53:74-85
>  @ ./src/components/App.js 22:0-49 136:43-56
>  @ ./src/index.js 15:0-35 31:36-39
> 
> WARNING in ./src/components/chat/MentionSuggestions.js 27:167-181
> export 'OnlineUserIcon' (imported as 'OnlineUserIcon') was not found in './UserStatusIcons' (possible exports: UserStatusIcon)
>  @ ./src/components/chat/AutoSuggest.js 5:0-54 18:130-148
>  @ ./src/components/chat/ChatContainer.js 20:0-40 53:74-85
>  @ ./src/components/App.js 22:0-49 136:43-56
>  @ ./src/index.js 15:0-35 31:36-39
> 
> WARNING in ./src/components/chat/MentionSuggestions.js 27:224-239
> export 'OfflineUserIcon' (imported as 'OfflineUserIcon') was not found in './UserStatusIcons' (possible exports: UserStatusIcon)
>  @ ./src/components/chat/AutoSuggest.js 5:0-54 18:130-148
>  @ ./src/components/chat/ChatContainer.js 20:0-40 53:74-85
>  @ ./src/components/App.js 22:0-49 136:43-56
>  @ ./src/index.js 15:0-35 31:36-39
> ```
> 
> Beyond that this looks pretty great to merge!

That's an existing problem. It's not caused by this branch. The new build is just superior in reporting issues like that. :)

I'm going to file an issue and fix the actual bug. Since it wasn't introduced here, I'd rather it didn't hold this back.

**Comment by elimisteve** [comment]
Merged! 🎉 Great job, @jimmcgaw 👌

**Pull request merged**